### PR TITLE
modules for hardware reservations

### LINKED ISSUE #101: Device additional features
- [ ] hw reservations

### PR DESCRIPTION
**Pull request by t0mk**
This PR adds modules for hw reservation. It also fixes hw ressevation support in `metal_device` .

There is a test, but it's (clumsily) disabled by default. It can be used locally to tests the modules if user has own hw reservation, and maybe it can be enabled in CI if we get a hw reservation to test. That would be further complicated by hw reservation "cooldown" period after a device is deleted.

This doesn't yet **move** the hardware reservation, as requested in #102 , I need to think a bit how to implement the move.

fixes #101

### CODE REVIEW

**Comment by t0mk** [comment]
equinix-metal 0.2.1 was released and this is ready for review

File: tests/integration/targets/metal_device_hw_res/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        test_prefix: "{{ test_resource_name_prefix }}-{{ unique_id }}"
+    - set_fact:
+        test_os: ubuntu_22_04
+    - set_fact:
+        test_plan: c3.small.x86
+    - set_fact:
+        wait_seconds: 1200
+    # set these two facts if you want to test the hw reservation resources
+    #- set_fact:
+    #    hardware_reservation_id: "82323c08-a7f5-4e09-8b34-634e82e527c1"
```
[the remark this answers is not in the document]
Comment (t0mk) [reply tests/integration/targets/metal_device_hw_res/tasks/main.yml:32]: @ctreatma I see your point. `--extra-vars` are not allowed in `ansible-test`, but variables can be defined in integration-config (https://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html#integration-config-yml) just like the API token. `integration-config.yml` is generated in Makefile, so if we use it, the tests can't be run directly with `ansible-test integration <test_name>`, but with `make`, for example, `make test_target=metal_project test`.

With the latest changes, the metal_device_hardware_reservation test will siletnly succeed if envvars
METAL_HARDWARE_RESERVATION_ID and METAL_HARDWARE_RESERVATION_PROJECT_ID
are unset. If they are set, the test will try to create the device in the resrevatoin. There's no need for editing the test yaml file.

I implemented it in the latest commit. I had to change the test run instructions in README.

What do you think?

[the remark this answers is not in the document]
Comment (ctreatma) [reply tests/integration/targets/metal_device_hw_res/tasks/main.yml:32]: I think that's a good approach.  Do we need to keep both tests around or can the fact-based one be removed?

[the remark this answers is not in the document]
Comment (t0mk) [reply tests/integration/targets/metal_device_hw_res/tasks/main.yml:32]: @ctreatma The fact-based approach was removed. Now the hardware reservation test runs if the two envvars are set, and it's skipped if they're unset.

**Comment by t0mk** [comment]
@ctreatma if you think all looks good, please approve this PR

**Review by ctreatma — APPROVED** [review]
Local test run passed.  🚀

**Pull request merged**